Add thread-safe Refresh Token Offset cache

AlexanderSher opened the pull request.

AlexanderSher committed:
Add thread-safe Refresh Token Offset cache

AlexanderSher commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -120,30 +75,102 @@ public async Task CachesHeaderValue()
-            var policy = new BearerTokenAuthenticationPolicy(credentialsMock.Object, "scope");
-            MockTransport transport = CreateMockTransport();
+        [Test]
+        public async Task BearerTokenAuthenticationPolicy_GatedConcurrentCalls()
```
More tests will follow

AlexanderSher replied:
Tests are added, code coverage for `BearerTokenAuthenticationPolicy` is 99% now

AlexanderSher commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -4,22 +4,30 @@
     /// <summary>
     /// A policy that sends an <see cref="AccessToken"/> provided by a <see cref="TokenCredential"/> as an Authentication header.
     /// </summary>
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable
```
`SemaphoreSlim` is required to be disposed, need to check how expensive it is to recreate new instance every time it is needed.

AlexanderSher replied:
Fixed

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +113,76 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                        ResetSemaphore(_headerSemaphore);
+                        return (null, null);
+                    case TokenState.Pending:
+                        _headerSemaphore ??= new SemaphoreSlim(0, 1);
```
Using convenience operators list ??= here and the ternary operator below can obscure synchronization issues in code which is intended to be thread safe, as what appears to be one statement is actually not executed atomically. I realize that this code is all under a lock so it's not really an issue, but if the code might be refactored in the future perhaps it's worth being explicit?

AlexanderSher replied:
`_headerSemaphore` should never be accessed outside of the lock. But maybe it will go away cause I'm thinking about replacing it with a different primitive.

AlexanderSher replied:
Extracted this logic into separate type, `AccessTokenCache`.

schaabs commented on sdk/core/Azure.Core/src/TokenCredential.cs:
```diff
@@ -11,6 +12,27 @@ namespace Azure.Core
+        /// attempts will be made to refresh the token. By default this will occur two minutes prior to the expiry of the token.
+        /// </summary>
+        /// <returns>The duration value representing the amount of time to subtract from the token expiry time.</returns>
+        protected internal TimeSpan TokenRefreshOffset { get; }
```
I was originally thinking this property could be virtual and the default implementation would return the default value. This would avoid the need for another protected constructor, and possibly provide a more consistent mocking story. Thoughts?

AlexanderSher replied:
The problem with `default` value is that default for `TimeSpan` is 2 minutes. We either have to make default value `null` and make the property nullable, or override default value internally, which isn't obvious for user.

schaabs replied:
I didn't mean default as in the C# keyword. I meant the default value of two minutes. Sorry for the ambiguity.

AlexanderSher replied:
Property is now `public virtual` and can be overridden in derived types

schaabs commented on sdk/core/Azure.Core/src/TokenCredential.cs:
```diff
@@ -11,6 +12,27 @@ namespace Azure.Core
+        /// attempts will be made to refresh the token. By default this will occur two minutes prior to the expiry of the token.
+        /// </summary>
+        /// <returns>The duration value representing the amount of time to subtract from the token expiry time.</returns>
+        protected internal TimeSpan TokenRefreshOffset { get; }
```
I think we had planned to make this return a structure or class `TokenRefreshOptions` rather than just a `TimeSpan` so it will be more forward compatible if we add more configuration around token refresh. This is in the java PR of this feature [here](https://github.com/Azure/azure-sdk-for-java/pull/12616/files#diff-1116664af60b3e486890e3b0312ba4aa)

AlexanderSher replied:
`TokenCredentialOptions` is defined in `Azure.Identity` while `TokenCredential` is defined in `Azure.Core`. I thought it would be a breaking change to move it.

schaabs replied:
I wasn't saying the type would be `TokenCredentialOptions` but a new type `TokenRefreshOptions`

AlexanderSher replied:
So you want a struct that currently has one property `TokenRefreshOffset`, correct? Should we then add a public setter to the property?

AlexanderSher replied:
Fixed

AlexanderSher committed:
- Introduce TokenRefreshOptions
- Use WithCancellationAwaitable<T> instead of SemaphoreSlim

AlexanderSher commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -8,12 +8,30 @@
+            return task.GetAwaiter().GetResult();
+        }
+
+        public static WithCancellationAwaitable<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
```
@pakrym , can you please review this change?

pakrym replied:
What's the benefit of having this method? Avoiding ThrowIfCancellationRequested before the method call?

AlexanderSher replied:
This is for scenario when a single task can be awaited several times, each with its own `CancellationToken`.

pakrym replied:
I still don't understand, in that scenario we would just check every token we need to check explicitly.

This extension method makes it look like the task would be actually cancelled when cancellation token is which is not true and might be confusing to consumers.

AlexanderSher replied:
Good point. I can rename the method into `AwaitWithCancellation` or `ConfigureAwaitWithCancellation`.
Checking isn't enough cause at the moment when `await` is reached it can be that neither task is completed nor cancellation token is canceled.

pakrym replied:
AwaitWithCancellation sounds good.

AlexanderSher replied:
Fixed

pakrym commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -133,5 +151,69 @@ private static void VerifyTaskCompleted(bool isCompleted)
+            public WithCancellationAwaiter<T> GetAwaiter() => new WithCancellationAwaiter<T>(_awaitable.GetAwaiter(), _cancellationToken);
+        }
+
+        public readonly struct WithCancellationAwaiter<T> : ICriticalNotifyCompletion
```
Why don't we implement this via an async method that takes a task?

AlexanderSher replied:
Not sure I understand your question. Async method that takes a task is `WithCancellation` above.

pakrym replied:
I mean why do we need a custom awaiter to get this feature.

Wouldn't the following be enough:
```
public static ValueTask<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
{
cancellationToken.ThrowIfCancellationRequested();
var result = await task;
cancellationToken.ThrowIfCancellationRequested();
return result;
}
``

AlexanderSher replied:
If `cancellationToken` is canceled after `await` is started but before `task` is completed, we shouldn't wait for that task, we should continue with throwing `OperationCanceledException`.

pakrym commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -8,12 +8,30 @@
+                // ignore exception here to rethrow it later
+            }
+
+            return task.GetAwaiter().GetResult();
```
This wouldn't this block even if cancellationToken is cancelled?

AlexanderSher replied:
Great catch! Fixed.

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +106,103 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                    switch (_tokenState) {
+                        case TokenState.Pending:
+                            _pendingTcs ??= new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+                            return (null, _pendingTcs.Task);
```
Who completes this TCS?

AlexanderSher replied:
`UpdateTokenState()`: https://github.com/Azure/azure-sdk-for-net/pull/13087/files#diff-8c190456ad69469f8f9e1ba176df1b68R153

AlexanderSher committed:
Fix build

AlexanderSher committed:
- Add tests
- Fix error with `cancellationToken` swallowed in `WaitWithCancellation`

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -64,19 +65,36 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                {
+                    headerValue = async
+                        ? await pending.WithCancellation(message.CancellationToken)
+                        : pending.WaitWithCancellation(message.CancellationToken);
```
BTW, this is an easy way to get a thread pool during peak loads.

AlexanderSher replied:
I'm not sure I understand how is it different from any other synchronous call that holds a thread in thread pool that we have.

pakrym replied:
I guess you are right it doesn't make things worse. It just looks like a chokepoint similar to the ones that I've seen causing issues in the past.

pakrym commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -133,5 +157,69 @@ private static void VerifyTaskCompleted(bool isCompleted)
+                Action reference = originalContinuation;
+                return () =>
+                {
+                    Action continuation = Interlocked.Exchange(ref reference, null);
```
Why do we need `Interlocked.Exchange` here? To prevent double registration?

pakrym replied:
I see, it's so we don't run continuation twice.

pakrym replied:
I'm not sure it works though.

In cancellation scenario both `originalContinuation` would be invoked first and if Task completes too the lambda would run next invoking `originalContinuation` again.

AlexanderSher replied:
Here is the code generated for [this closure](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQBUAWATgKYCGwAlgHYDmA3MmpjgEoCuVALhQLbHYBhAPY8ADhQA2xQgGVpANwoBjYgGcGSRgGZ0qzoTZLO6AOoVO+AaSoqJE0tyFUAggHdS50gCMpAHlwAfOgA3sjo4eiihBTyDsToJOROEgCe6FY2xHYOFE64QgDWxFToAPpK1rb2jlT5RVQaEZHRsZzxicDJacJUAGYUNGwkwFAAbG4enN5+gYJO/YPDYxPm0mVTqgUrbYQaYRFRMXHMMOmJbT3cVGw5TgAU1CfoQtE01KQSl9Q3NQCU++FQkgmk0KL10HcAITlSpZaq5WqFYqCawAIWIGVsxGA6AgEHWpE222k2AAkqphGIpG1gP9gSDAQCGREoAB2Z6vd6fJxXH4IxrMgC+yCZTUxcNuiPqLGIbz0hElCVlFHlioAvGUKplsjU6siZXKdncXgMuV9rpLfgKGVhTiRetJiip0BqTW8qB9zXynNaQWyIb8XQFRREgcy/XBTkoed91ehSVxpBIhEoisBsABRAAeSnw1hoxDu9qVDpImQANOhrnYrSGGWCIdGuLGauhIRrqxI6eGmmGewySIaFTVsAARFWiISqQu1+n9ppN3mSu6z+foYVzkGC30bwVAA=):
```cs
private Action CreateContinuation([In] [IsReadOnly] ref Action originalContinuation)
{
    <>c__DisplayClass2_0 <>c__DisplayClass2_ = new <>c__DisplayClass2_0();
    if (!_cancellationToken.CanBeCanceled)
    {
        return originalContinuation;
    }
    <>c__DisplayClass2_.registration = _cancellationToken.Register(originalContinuation);
    <>c__DisplayClass2_.reference = originalContinuation;
    return new Action(<>c__DisplayClass2_.<CreateContinuation>b__0);
}
```
```cs
private sealed class <>c__DisplayClass2_0
{
    public Action reference;
    public CancellationTokenRegistration registration;
    internal void <CreateContinuation>b__0()
    {
        Action action = Interlocked.Exchange(ref reference, null);
        if (action != null)
        {
            registration.Dispose();
            action();
         }
    }
}
```
`originalContinuation` is assigned to the `reference`, and then `reference` is replaced with `null`, so it is called only once. `registration.Dispose()` can be called twice but that is ok.

pakrym replied:
> originalContinuation is assigned to the reference, and then reference is replaced with null

Yes, but you are also using the `originalContinuation` as the cancellation token callback which would fire unconditionally (it's never replaced by null). Cancellation token callback doesn't clear the `reference` so continuation would still run the `originalContinuation` once.

What am I missing?

AlexanderSher replied:
No, you're actually right. That's a great catch. Fixed and added tests.

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +106,103 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                    }
+                    else
+                    {
+                        pendingTcs.SetException(exception ?? new InvalidOperationException());
```
Let's add a message to this exception. In unlikely case of this scenario happening it would help us debug.

AlexanderSher replied:
Fixed

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +106,103 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                return headerValue;
+            }
+
+            public (string? headerValue, Task<string>? pendingTask) UpdateTokenState()
```
Why don't you return a ValueTask<string> here?

AlexanderSher replied:
You mean to use the `ManualResetValueTaskSourceCore` instead of `TaskCompletionSource`?

pakrym replied:
No, don't return the tuple and handle actual value retrieval inside AccessTokenCache

AlexanderSher replied:
Done

pakrym commented on sdk/core/Azure.Core/src/TokenCredential.cs:
```diff
@@ -11,6 +12,13 @@ namespace Azure.Core
+        /// attempts will be made to refresh the token. By default this will occur two minutes prior to the expiry of the token.
+        /// </summary>
+        /// <returns>The duration value representing the amount of time to subtract from the token expiry time.</returns>
+        public virtual TokenRefreshOptions TokenRefreshOffset { get; } = new TokenRefreshOptions(TimeSpan.FromMinutes(2));
```
Should this property name be `RefrestOptions`?

pakrym replied:
Can't we control the same thing by changing ExpiresOn ? https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/AccessToken.cs#L32

AlexanderSher replied:
No, we need the value of `ExpireOn` as well so that while one request refreshes the token the other requests can use the old one while it is valid.

pakrym commented:
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/7442

AlexanderSher committed:
Renames

AlexanderSher committed:
Remove _taskAwaiter.IsCompleted check to avoid concurrent calls of `originalContinuation`

AlexanderSher committed:
Add text to the exception

AlexanderSher committed:
Fix the issue with double call of the continuation

pakrym commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -133,5 +157,80 @@ private static void VerifyTaskCompleted(bool isCompleted)
 #pragma warning restore AZC0107 // Do not call public asynchronous method in synchronous scope.
         }
+
+        public readonly struct WithCancellationAwaitable<T>
```
I'm worried about this implementation being a bit complicated.

Can we go with something simpler like:
            
``` C#
var tcs = new TaskCompletionSource<T>();
_registration = cancellationToken.Register(() => tcs.TrySetCanceled(cancellationToken));

await Task.WhenAny(tcs.Task, task)
```

??

AlexanderSher replied:
- With a switch to `ValueTask`, `Task.WhenAny` can't be used directly and will require `AsTask()` to be called
- `Task.WhenAny(tcs.Task, task).Unwrap().ConfigureAwait(false)` is much slower (because of all the task continuation overhead) and produces a lot of garbage in memory. Awaiter in the worst case requires only two objects - continuation wrapper and delegate.

pakrym replied:
Currently I'm worried about future supportability of this code.

It doesn't like any of the current uses is perf sensitive.

AlexanderSher replied:
I don't know how often `BearerTokenAuthenticationPolicy.ProcessAsync` can be called, however there are at least 2 more places in `Azure.Identity` where this method will be useful, so I'd like to make it publicly available. Which means we won't know where and how often it will be used.

pakrym commented:
Was this API approved?

AlexanderSher committed:
- Use ValueTask instead of Tuple
- Move Wait method from TaskExtensions to BearerTokenAuthenticationPolicy (this method is blocking so shouldn't be publicly available).
- Import DateTimeOffsetHelper from dotnet/runtime to get precise DateTimeOffset in .net framework

pakrym commented on sdk/core/Azure.Core/src/Internal/DateTimeOffsetHelpers.netfx.cs:
```diff
@@ -0,0 +1,77 @@
+        /// </summary>
+        private static DateTimeOffset GetUtcNowImplementation()
+        {
+            // DateTime.UtcNow accuracy on .NET Framework is ~16ms, this method
```
Why do we need more than 16ms accuracy ?

AlexanderSher replied:
This has been causing one of the tests to fail from time to time. Also, I'll need this to be able to correctly set start time of the activity in another changeset.

pakrym replied:
Can we fix the test instead? Seems like a lot of extra production code to pull in to fix the test.

AlexanderSher replied:
Fixed

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -87,6 +99,116 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                    task.Wait(cancellationToken);
+                }
+                catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+                return task.EnsureCompleted();
```
EnsureCompleted would wait for task to finish even if `cancellationToken` have fired, right?

AlexanderSher replied:
No, we won't reach this method cause `task.Wait(cancellationToken);` will throw `OperationCanceledException`, not `AggregateException`

pakrym commented on sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:
```diff
@@ -133,5 +137,80 @@ private static void VerifyTaskCompleted(bool isCompleted)
+            public bool IsCompleted => _taskAwaiter.IsCompleted || _cancellationToken.IsCancellationRequested;
+
+            public void OnCompleted(Action continuation)
+                => _taskAwaiter.OnCompleted(WrapContinuation(continuation));
```
Do we need to capture sync and execution context here and restore it before calling the continutation?

AlexanderSher committed:
- Remove DateTimeOffsetHelpers
- Update the test

schaabs commented on sdk/core/Azure.Core/src/TokenCredential.cs:
```diff
@@ -11,6 +12,13 @@ namespace Azure.Core
+        /// attempts will be made to refresh the token. By default this will occur two minutes prior to the expiry of the token.
+        /// </summary>
+        /// <returns>The duration value representing the amount of time to subtract from the token expiry time.</returns>
+        public virtual TokenRefreshOptions RefreshOffset { get; } = new TokenRefreshOptions(TimeSpan.FromMinutes(2));
```
This should be named `RefreshOptions` rather than `RefreshOffset`

AlexanderSher replied:
Fixed

AlexanderSher committed:
Rename RefreshOffset property to RefreshOptions

pakrym commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -120,30 +75,264 @@ public async Task CachesHeaderValue()
+            }, IsAsync);
+
+            var policy = new BearerTokenAuthenticationPolicy(credential, "scope");
+            MockTransport transport = CreateMockTransport(Enumerable.Range(0, 100).Select(i => new MockResponse(200)).ToArray());
```
You can use the constructor that takes factory func here.

AlexanderSher replied:
Fixed

AlexanderSher committed:
Use factory in tests

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -87,6 +99,116 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                        case TokenState.Valid:
+                        case TokenState.AboutToExpire:
+                            return new ValueTask<string?>(_headerValue);
+                        default:
```
What about TokenState.Invalid?

AlexanderSher replied:
Handled above by:
```cs
if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
{
    _tokenState = TokenState.Pending;
    _headerValue = null;
    return new ValueTask<string?>();
}
```

If state is `Invalid` and `DateTimeOffset.UtcNow < _expiresOn` then something went wrong.

AlexanderSher committed:
Refactoring

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -99,19 +80,47 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+        }
 
-            static string? WaitSynchronously(ValueTask<string?> headerValueTask, CancellationToken cancellationToken)
+        private async ValueTask<string?> GetHeaderValueFromCredentialAsync(HttpMessage message, bool async) {
```
nit formatting.

pakrym commented:
@KrzysztofCwalina can you sign off on this please?

AlexanderSher committed:
Remove public API changes

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +81,146 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            return task.EnsureCompleted();
+        }
+
+        private class AccessTokenCache
```
I would extract it when we implement caching in one of those policies.

schaabs replied:
At least the `ChallengeBasedAuthenticationPolicy` in key vault currently implements caching, in the same way that `BearerTokenAuthenticationPolicy` does. But we can wait until we update that to figure out how to refactor this into a shared type.

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -64,19 +64,12 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                message.Request.SetHeader(HttpHeader.Names.Authorization, headerValue);
             }
```
What's the scenario where we'll get here with null header value? I think we should be throwing in some other place with an actual reason we weren't able to get a header value.

schaabs replied:
Most likely we would have thrown an AuthenticationFailedException, so `headerValue` being null is probably a case which is not reachable under normal conditions. Certainly it wouldn't be expected.

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +81,146 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            }
+            catch (Exception e)
+            {
+                _accessTokenCache.ResetToken(e);
```
How often would we try to refresh the background token? Do we want to try and refresh in the background without blocking the current request at all?

AlexanderSher replied:
This code is unreachable when token hasn't expired. I've added test for that case.

schaabs replied:
@pakrym this is what we decided to do in JS. Essentially we queue the refresh when the first request comes in after we've reached the refresh buffer, but we don't block that request. We could use the same functionality.

AlexanderSher replied:
Added another test, fixed the issue

pakrym replied:
The current implementation still does token refresh inline with a request or am I missing something?

AlexanderSher replied:
I think it is. Previously, if request has failed, existing token will be erased even if it isn't expired. With this change, it will be reset only if it has already expired, otherwise cache will be returned in the `TokenState.Valid` so that next request can try to refresh token again:

```cs
public void ResetTokenIfExpired(Exception exception)
{
string? headerValue = default;
    TaskCompletionSource<string?>? pendingTcs;

    lock (_syncObj)
    {
        if (DateTimeOffset.UtcNow >= _expiresOn)
        {
            _headerValue = default;
            _refreshOn = default;
            _expiresOn = default;
            _tokenState = TokenState.Invalid;

            pendingTcs = _pendingTcs;
            _pendingTcs = null;
        }
        else if (_tokenState == TokenState.AboutToExpire)
        {
            _tokenState = TokenState.Valid;
            if (_pendingTcs != default)
            {
                exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
            }
        }
    }

    if (headerValue != null)
    {
        pendingTcs?.SetResult(headerValue);
    }
    else
    {
        pendingTcs?.SetException(exception);
    }
}
```

Here is the situation we handle: Two options are possible: If at the time of request failure

pakrym replied:
Yes, but trying to get a new token would still happen inline with sending the current request not on a background thread.

AlexanderSher replied:
Fixed

AlexanderSher committed:
- Added delays into tests
- Added test for failure after expire

heaths commented on sdk/core/Azure.Core.TestFramework/src/MockTransport.cs:
```diff
@@ -63,7 +70,10 @@ private async Task ProcessCore(HttpMessage message)
                 throw new InvalidOperationException("the request is not compatible with the transport");
 
-            Requests.Add(request);
+            lock (_syncObj)
```
To note, this should also fix #13195

AlexanderSher committed:
- Address PR comments
- Fix BearerTokenAuthenticationPolicy_TokenAlmostExpired
- Add BearerTokenAuthenticationPolicy_TokenAlmostExpiredThenFailed

AlexanderSher committed:
Upload missing change

AlexanderSher committed:
- Change behavior for refreshing token to desired
- Fix tests to validate expected results

AlexanderSher commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -64,21 +64,18 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            else if (scheduleRefreshToken)
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                _ = Task.Run(() => GetHeaderValueFromCredentialAsync(message, true));
```
@pakrym Do we have any better way to handle this task than make it just hanging in the void?

pakrym replied:
What happens if this task fails? Would we ever try again before the token fully expires or when it does expire?

AlexanderSher replied:
If it fails, `_accessTokenCache.ResetTokenIfExpired` will be called. If token has expired while `TokenCredential.GetTokenAsync` was executed, `AccessTokenCache` will be set into `Invalid` state. If it is still not expired, it will be put back into `Valid` state so that next request can trigger another update attempt.

AlexanderSher committed:
Pass `async` instead of using `true` cause `TokenCredential` implementation may not have `async` implementation

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                        return (default, default, false);
+                    }
+
+                    if (DateTimeOffset.UtcNow >= _refreshOn && _tokenState == TokenState.Valid)
```
This needs to be updated to ensure we don't refresh more often than every 30 seconds, if the token is still not expired.

AlexanderSher replied:
Fixed, changed test

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                    {
+                        _tokenState = TokenState.Pending;
+                        _headerValue = default;
+                        return (default, default, false);
```
I'm not sure why we return false in this case. Do we not want to schedule a refresh if the token is expired?

AlexanderSher replied:
No, we don't want to **schedule** a refresh, we want to refresh immediately.

schaabs replied:
Perhaps an xml comment on the method that describes the meaning and usage of the data returned.

AlexanderSher replied:
Done

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                        _tokenState = TokenState.Valid;
+                        if (_pendingTcs != default)
+                        {
+                            exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
```
Should you do this check and throw before you reset the state to valid? Otherwise is the cache in an inconsistent state if the exception is handled?

AlexanderSher replied:
No, but I need to set `_pendingTcs` to null.  We should get out of the `TokenState.AboutToExpire` state anyway.

AlexanderSher replied:
Fixed

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                    }
+                    else if (_tokenState == TokenState.AboutToExpire)
+                    {
+                        _tokenState = TokenState.Valid;
```
Why does an Exception when the state is `AboutToExpire` result in setting the state to `Valid`?

AlexanderSher replied:
Since token hasn't expired, we want to return back to `Valid` state so that next call can make another attempt

schaabs replied:
I think we need a detailed comment somewhere on this class which describes the meaning of each state and how they are used.

AlexanderSher replied:
Refactored this piece. State field is replaced with `_isGettingToken` which is `true` while `GetToken` is running.

AlexanderSher replied:
States are removed in new implementation

AlexanderSher committed:
- Added refresh delay support
- Fixed tests
- Refactored for better readability

AlexanderSher committed:
Added comment to GetHeaderValue

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -64,21 +64,18 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+            else if (refreshTokenInBackground)
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                _ = Task.Run(() => GetHeaderValueFromCredentialAsync(message, async));
```
Can we handle this in GetHeaderValueAsync?

AlexanderSher replied:
Not really. It either requires to pass a callback or a direct reference to the `TokenCredential`. I have renamed `refreshTokenInBackground` into `accessTokenAboutToExpire`.

AlexanderSher replied:
Resolved with new implementation

AlexanderSher committed:
Rename output tuple item

AlexanderSher committed:
Attempt to merge TokenCredential.GetToken into AccessTokenCache

AlexanderSher committed:
Another attempt to simplify code.

AlexanderSher committed:
- Reduce code under lock to one method
- More test coverage

AlexanderSher committed:
Remove explicit states from AccessTokenCache

AlexanderSher committed:
Fix typo

AlexanderSher committed:
Add logging

AlexanderSher committed:
increase timeout in tests

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -56,36 +51,167 @@ public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePol
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                lock (_syncObj)
+                {
+                    if (_infoTcs == null)
```
Can you please add a few comments about what conditions in this method mean logically?

AlexanderSher replied:
Done

pakrym commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -56,36 +51,167 @@ public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePol
-                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                try
+                {
+                    HeaderValueInfo newInfo = await GetHeaderValueFromCredentialAsync(httpMessage, async);
```
An interesting side effect of this implementation is that we would use current request's cancellation token for a background refresh.

AlexanderSher replied:
Yeah, but we don't have any better `CancellationToken`

pakrym replied:
Is it better not to have any? Or have one with a baked in timeout (1 minute)?

Because current behavior is bad any way you look at it:
- If customer cancels the request they would cancel an unrelated token refresh too.
- If customer doesn't cancel background refresh would run unattended.

AlexanderSher replied:
Will be using `_tokenRefreshRetryDelay` for timeout, and run next retry without delay if this has failed on timeout

AlexanderSher committed:
Address CR comments

AlexanderSher committed:
Updated wrong error message

schaabs commented on sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:
```diff
@@ -13,35 +15,28 @@ namespace Azure.Core.Pipeline
-        {
+            : this(credential, scopes, TimeSpan.FromMinutes(5), TimeSpan.FromSeconds(30)) { }
+
+        internal BearerTokenAuthenticationPolicy(TokenCredential credential, IEnumerable<string> scopes, TimeSpan tokenRefreshOffset, TimeSpan tokenRefreshRetryTimeout) {
```
Can we rename `tokenRefreshRetryTImeout` everywhere? I know it's all internal, but I think the name doesn't really express what it is used for and so it's more difficult to undertand the code. I think `retryDelay` or `retryRate` would be more discriptive. It would need to be changed bot there and in `AccessTokenCache`.

AlexanderSher replied:
Fixed

AlexanderSher committed:
Rename

AlexanderSher commented on sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs:
```diff
@@ -32,6 +33,12 @@ internal sealed class AzureCoreEventSource : EventSource
 
         public static AzureCoreEventSource Singleton { get; } = new AzureCoreEventSource();
 
+        [Event(BackgroundRefreshFailedEvent, Level = EventLevel.Informational, Message = "Authentication request [{0}] exception {1}")]
```
Fixed

AlexanderSher committed:
Add checks in test

AlexanderSher committed:
Set offset time before synchronization

AlexanderSher committed:
Increase token expiration period to 1 second
